Add Loki Promtail (#688)
netgroup-polito/CrownLabs: infrastructure/monitoring/loki/README.md

# Loki

Grafana Loki is a set of elements that can be composed into a fully featured logging stack.

> Loki is built around the idea of only indexing metadata about your logs with labels (just like Prometheus labels). The logging data is packed into chunks and stored in object stores like S3. A small index and highly compressed chunks simplifies the operation and significantly lowers the cost of Loki.

Reviewer: ```suggestion
Loki is built around the idea of indexing logs metadata only by means of labels (just like Prometheus labels).
The actual logs data is packed into chunks and stored in object stores like S3. 
A small index and highly compressed chunks simplify the operations and significantly lower the cost of Loki.
```

Nit. Create a new line for each sentence, to simplify versioning.

> In our cluster we rely on Rook Ceph as a distributed storage provider. For a more detailed explanation of the object storage creation the official guide is here https://rook.io/docs/rook/v1.3/ceph-object.html, in this guide we present only the main steps.

Reviewer: ```suggestion
In our cluster we rely on Rook Ceph as a distributed storage provider. 
For a more detailed explanation about the object storage creation you can rely on the [official guide](https://rook.io/docs/rook/v1.3/ceph-object.html).
```

Rook 1.3 seems pretty old to me.

## Prerequisites

> Helm must be installed to use the charts value. Please follow the [Helm installing guide](https://helm.sh/docs/intro/install/) before getting starting.

Reviewer: ```suggestion
Helm must be installed to deploy the chart. Please follow the [Helm installing guide](https://helm.sh/docs/intro/install/) before getting starting.
```

After Helm is ready, add the Grafana repo as follow:

```helm
helm repo add grafana https://grafana.github.io/helm-charts
helm repo update
```
To see the Grafan charts, you can run:

```helm
helm search repo grafana
```

## Installing

> To install the chart with the release name loki and apply the configuration of values.yaml:

Reviewer: ```suggestion
To install the chart with the release name `loki` and apply the configuration specified by the `values.yaml` file, it is possible to proceed as follows:
```

```helm

> helm upgrade --install loki -f < YOUR_PATH >/values.yaml grafana/loki

Reviewer: Same here about the additional flags.

```

> The [values.yaml](./values.yaml) is the file value for Loki configuration.

Reviewer: Same here about adding a bit of introduction.

```yaml
  ingester:
    ...
    wal:
      dir: /data/loki/wal
    lifecycler:
      ring:
        kvstore:
          store: memberlist
        replication_factor: 3
    ...
  memberlist:
    abort_if_cluster_join_fails: false

    join_members:
    - loki-headless
```

> The <b>store</b> is the backend storage to use for the ring, it can be **inmemory**, **memberlist**, **cunsul** and **etcd**. 

Reviewer: ```suggestion
The **store** is the backend storage to use for the ring, it can be **inmemory**, **memberlist**, **cunsul** and **etcd**. 
```

In this case, with **memberlist** the information is shared between the nodes, in this way it spreads faster, it also reduces collisions and detect node failure.

```yaml
  schema_config:
    configs:
    - from: 2020-10-24
      store: boltdb-shipper
      object_store: s3 #filesystem
      schema: v11
      index:
        prefix: index_
        period: 24h
    ...
  storage_config:
    boltdb_shipper:
      active_index_directory: /data/loki/boltdb-shipper-active
      cache_location: /data/loki/boltdb-shipper-cache
      cache_ttl: 24h         # Can be increased for faster performance over longer query periods, uses more disk space
      shared_store: s3 #filesystem
    aws:
      s3: s3://<ACCESS_KEY>:<SECRET_ACCESS_KEY>@BUCKET_HOST/BUCKET_NAME
      s3forcepathstyle: true
    filesystem:
      directory: /data/loki/chunks
    ...
  compactor:
    working_directory: /data/loki/boltdb-shipper-compactor
    shared_store: s3 #filesystem
```

> You can choose where you want to store the data, in filesystem or in s3. For s3 you must configure s3 as indicated. Some examples of Loki configuration are [here](https://grafana.com/docs/loki/latest/configuration/examples/#loki-configuration-examples).

Reviewer: ```suggestion
You can choose where you want to store the data, either in the local filesystem (PVC) or in a S3 cobject store. For the latter, you must configure the S3 bucket parameters as indicated. Some examples of Loki configurations are also available [here](https://grafana.com/docs/loki/latest/configuration/examples/#loki-configuration-examples).
```

*NOTE: some buckets, for example AWS, want the region before the BUCKET_HOST. So the s3 is s3://<ACCESS_KEY>:<SECRET_ACCESS_KEY>@REGION_BUKET.BUCKET_HOST/BUCKET_NAME*.

### Bibliography

https://github.com/grafana/helm-charts/tree/main/charts/loki <br>
https://grafana.com/docs/loki/latest/configuration/ <br>

> https://github.com/hashicorp/memberlist

Reviewer: Same here about the bibliography style
